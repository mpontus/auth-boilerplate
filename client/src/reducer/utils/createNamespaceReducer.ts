import { Action, Reducer } from "redux";

type Namespace<K extends string, V> = Partial<Record<K, V>>;

type HigherOrderReducer<S, A extends Action<any>> = (
  action: A
) => (state: S, action: A) => S;

export const createNamespaceReducer = <K extends string, V, A extends Action>(
  fn: HigherOrderReducer<Namespace<K, V>, A>
): Reducer<Namespace<K, V>, A> => (state = {}, action) =>
  fn(action)(state, action);
