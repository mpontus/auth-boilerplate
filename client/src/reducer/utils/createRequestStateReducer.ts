import { Action, Reducer } from "redux";

interface RequestState<ErrorType> {
  loading: boolean;
  success: boolean;
  error?: ErrorType;
}

const initialState = {
  loading: false,
  success: false
};

/**
 * Reusable reducer constructor for managing request state
 */
export const createRequestStateReducer = <
  ActionType extends Action<any>,
  ErrorType extends Error
>(
  reducer: (
    state: RequestState<ErrorType>,
    action: ActionType
  ) => RequestState<ErrorType>
): Reducer<RequestState<ErrorType>, ActionType> => (
  state = initialState,
  action
) => reducer(state, action);
