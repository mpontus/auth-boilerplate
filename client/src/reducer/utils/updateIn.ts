import { Action, Reducer } from "redux";

/**
 * Return a function which updates given key in an object, using the
 * provided updater function.
 */
export const updateIn = <
  K extends keyof R,
  A extends Action,
  V,
  R = Record<K, V>
>(
  key: K,
  updater: Reducer<R[K], A>
) => (state: R, action: A): R =>
  Object.assign({}, state, { [key]: updater(state[key], action) });
