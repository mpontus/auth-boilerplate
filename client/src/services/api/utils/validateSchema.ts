import * as t from "io-ts";
import { PathReporter } from "io-ts/lib/PathReporter";

export const validateSchema = <TSchema extends t.Type<any>>(
  schema: TSchema,
  data: any
): t.TypeOf<TSchema> => {
  const result = schema.decode(data);

  if (result.isLeft()) {
    throw new Error(PathReporter.report(result).join("\n"));
  }

  return result.value;
};
