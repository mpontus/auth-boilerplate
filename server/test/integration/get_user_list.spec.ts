import supertest from 'supertest';
import { initApp } from '../utils/initApp';
import { resetDb } from '../utils/resetDb';

let expressApp: any;
let nestApp: any;

beforeAll(async () => {
  ({ expressApp, nestApp } = await initApp());
});

afterAll(() => nestApp.close());

beforeEach(resetDb);

describe('get user by id', () => {
  const seed = require('../seeds/many_users');

  beforeEach(() => seed.run());

  describe('when requested by admin', () => {
    const adminSeed = require('../seeds/authenticated_admin_user');

    beforeEach(() => adminSeed.run());

    describe('when pagination is omitted', () => {
      it('should succeed', async () => {
        const response = await supertest(expressApp)
          .get('/users')
          .set('Authorization', `Bearer ${adminSeed.token}`)
          .expect(200);

        expect(response.body).toMatchSnapshot();
      });
    });

    describe('when offset is provided', () => {
      it('should succeed', async () => {
        const response = await supertest(expressApp)
          .get('/users?skip=2')
          .set('Authorization', `Bearer ${adminSeed.token}`)
          .expect(200);

        expect(response.body).toMatchSnapshot();
      });
    });

    describe('when limit is provided', () => {
      it('should succeed', async () => {
        const response = await supertest(expressApp)
          .get('/users?take=2')
          .set('Authorization', `Bearer ${adminSeed.token}`)
          .expect(200);

        expect(response.body).toMatchSnapshot();
      });
    });

    describe('when offset is invalid', () => {
      it('should fail', async () => {
        const response = await supertest(expressApp)
          .get('/users?skip=foo')
          .set('Authorization', `Bearer ${adminSeed.token}`)
          .expect(400);

        expect(response.body).toMatchSnapshot();
      });
    });

    describe('when offset is negative', () => {
      it('should fail', async () => {
        const response = await supertest(expressApp)
          .get('/users?skip=-2')
          .set('Authorization', `Bearer ${adminSeed.token}`)
          .expect(400);

        expect(response.body).toMatchSnapshot();
      });
    });

    describe('when limit is invalid', () => {
      it('should fail', async () => {
        const response = await supertest(expressApp)
          .get('/users?take=foo')
          .set('Authorization', `Bearer ${adminSeed.token}`)
          .expect(400);

        expect(response.body).toMatchSnapshot();
      });
    });

    describe('when limit is negative', () => {
      it('should fail', async () => {
        const response = await supertest(expressApp)
          .get('/users?take=-2')
          .set('Authorization', `Bearer ${adminSeed.token}`)
          .expect(400);

        expect(response.body).toMatchSnapshot();
      });
    });

    describe('when limit is too large', () => {
      it('should fail', async () => {
        const response = await supertest(expressApp)
          .get('/users?take=90000')
          .set('Authorization', `Bearer ${adminSeed.token}`)
          .expect(400);

        expect(response.body).toMatchSnapshot();
      });
    });
  });

  describe('when requested by someone else', () => {
    const userSeed = require('../seeds/authenticated_user');

    beforeEach(() => userSeed.run());

    it('should fail', async () => {
      const response = await supertest(expressApp)
        .get(`/users`)
        .set('Authorization', `Bearer ${userSeed.token}`)
        .expect(401);

      expect(response.body).toMatchSnapshot();
    });
  });
});
