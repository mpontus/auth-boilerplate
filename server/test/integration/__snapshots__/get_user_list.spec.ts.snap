// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get user by id when requested by admin when limit is invalid should fail 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    Object {
      "children": Array [],
      "constraints": Object {
        "isPositive": "take must be a positive number",
        "max": "take must not be greater than 50",
      },
      "property": "take",
      "target": Object {
        "take": null,
      },
      "value": null,
    },
  ],
  "statusCode": 400,
}
`;

exports[`get user by id when requested by admin when limit is negative should fail 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    Object {
      "children": Array [],
      "constraints": Object {
        "isPositive": "take must be a positive number",
      },
      "property": "take",
      "target": Object {
        "take": -2,
      },
      "value": -2,
    },
  ],
  "statusCode": 400,
}
`;

exports[`get user by id when requested by admin when limit is provided should succeed 1`] = `
Object {
  "items": Array [
    Object {
      "id": "1020792911429632",
      "name": "Derek Blake",
    },
    Object {
      "id": "4433688441913344",
      "name": "Phoebe Sandoval",
    },
  ],
  "total": 32,
}
`;

exports[`get user by id when requested by admin when limit is too large should fail 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    Object {
      "children": Array [],
      "constraints": Object {
        "max": "take must not be greater than 50",
      },
      "property": "take",
      "target": Object {
        "take": 90000,
      },
      "value": 90000,
    },
  ],
  "statusCode": 400,
}
`;

exports[`get user by id when requested by admin when offset is invalid should fail 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    Object {
      "children": Array [],
      "constraints": Object {
        "isPositive": "skip must be a positive number",
      },
      "property": "skip",
      "target": Object {
        "skip": null,
      },
      "value": null,
    },
  ],
  "statusCode": 400,
}
`;

exports[`get user by id when requested by admin when offset is negative should fail 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    Object {
      "children": Array [],
      "constraints": Object {
        "isPositive": "skip must be a positive number",
      },
      "property": "skip",
      "target": Object {
        "skip": -2,
      },
      "value": -2,
    },
  ],
  "statusCode": 400,
}
`;

exports[`get user by id when requested by admin when offset is provided should succeed 1`] = `
Object {
  "items": Array [
    Object {
      "id": "7591496318451712",
      "name": "Della Hale",
    },
    Object {
      "id": "6661526800826368",
      "name": "Jeffery Munoz",
    },
    Object {
      "id": "5296032296468480",
      "name": "Norman Lane",
    },
    Object {
      "id": "2475574186475520",
      "name": "Louis Moran",
    },
    Object {
      "id": "810085049696256",
      "name": "Bobby Stevens",
    },
    Object {
      "id": "2406117441273856",
      "name": "Mina Wallace",
    },
    Object {
      "id": "654970728742912",
      "name": "Edna Davis",
    },
    Object {
      "id": "4309729140539392",
      "name": "Frank Watson",
    },
    Object {
      "id": "5548803264348160",
      "name": "Dorothy Burgess",
    },
    Object {
      "id": "2348784637968384",
      "name": "Olivia Delgado",
    },
  ],
  "total": 32,
}
`;

exports[`get user by id when requested by admin when pagination is omitted should succeed 1`] = `
Object {
  "items": Array [
    Object {
      "id": "1020792911429632",
      "name": "Derek Blake",
    },
    Object {
      "id": "4433688441913344",
      "name": "Phoebe Sandoval",
    },
    Object {
      "id": "7591496318451712",
      "name": "Della Hale",
    },
    Object {
      "id": "6661526800826368",
      "name": "Jeffery Munoz",
    },
    Object {
      "id": "5296032296468480",
      "name": "Norman Lane",
    },
    Object {
      "id": "2475574186475520",
      "name": "Louis Moran",
    },
    Object {
      "id": "810085049696256",
      "name": "Bobby Stevens",
    },
    Object {
      "id": "2406117441273856",
      "name": "Mina Wallace",
    },
    Object {
      "id": "654970728742912",
      "name": "Edna Davis",
    },
    Object {
      "id": "4309729140539392",
      "name": "Frank Watson",
    },
  ],
  "total": 32,
}
`;

exports[`get user by id when requested by someone else should fail 1`] = `
Object {
  "error": "Unauthorized",
  "statusCode": 401,
}
`;
